reactive/kubernetes_worker.py

```diff
@@ -242,6 +242,49 @@ def shutdown():
     service_stop('snap.kube-proxy.daemon')
 
 
+@when_not('config.changed.open-ports')
+def open_close_ports_if_needed():
+
+    # 1) Check which ports are open and save to already_opened
+    # opened_ports return format: "icmp" or "80/tcp"
+    already_opened = hookenv.opened_ports()
+    to_open = set(p.strip() for p in hookenv.config().get("open-ports","").split(','))
+    # first need to check if to_open has ports on right format, which are:
+    # PORT --> PORT is a number or icmp
+    # PORT/PROTOCOL --> PROTOCOL can only be tcp or udp
+    # PORT_1-PORT_2 --> both port codes must be numbers
+    # PORT_1-PORT_2/PROTOCOL --> both port codes must be numbers and protocol tcp or udp
+    for p in to_open:
+        port = p
+        if "/" in p:
+            port = p.split("/")[0]
+            if p.split("/")[1] != "tcp" and p.split("/")[1] != "udp":
+                # Log is for unit test purposes
+                hookenv.log("[ERROR] open-ports have wrong port format on: {}. Only tcp or udp allowed as protocols on PORT/PROTOCOL".format(p))
+                raise Exception("[ERROR] open-ports have wrong port format on: {}. Only tcp or udp allowed as protocols on PORT/PROTOCOL".format(p))
+        if not "-" in port:
+            if port != "icmp" and not port.isdigit():
+                # Log is for unit test purposes
+                hookenv.log("[ERROR] open-ports have wrong port format on: {}. PORT should be a number and format: PORT[/PROTOCOL]".format(p))
+                raise Exception("[ERROR] open-ports have wrong port format on: {}. PORT should be a number and format: PORT[/PROTOCOL]".format(p))
+        else:
+            if port != "icmp" and not (port.split("-")[0].isdigit() and port.split("-")[1].isdigit()):
+                # Log is for unit test purposes
+                hookenv.log("[ERROR] open-ports have wrong port format on: {}. Port should be a number and format: PORT_1-PORT_2[/PROTOCOL]".format(p))
+                raise Exception("[ERROR] open-ports have wrong port format on: {}. Port should be a number and format: PORT_1-PORT_2[/PROTOCOL]".format(p))
+    # assuming all those ports will be closed
+    for p in already_opened:
```

Comment:
Can we pluck out of already_opened as we go through to_open to prevent bouncing the ports?

Reply:
So, check my comment below.
(This can be very complex if we take ranges into consideration)